contracts/vesting/src/contract.rs

```diff
@@ -0,0 +1,603 @@
+            );
+            panic_with_error!(env, ContractError::NotAuthorized);
+        }
+
+        save_minter(
+            &env,
+            MinterInfo {
+                address: new_minter.clone(),
+                cap: get_minter(&env).cap,
+            },
+        );
+
+        env.events()
+            .publish(("Update minter", "Updated minter to: "), new_minter);
+    }
+
+    // TOOD: the only difference between this and  `transfer_token` is that here we do not check the vesting
+    fn send_tokens_to_contract(env: Env, sender: Address, contract: Address, amount: i128) {
+        if amount <= 0 {
+            log!(&env, "Vesting: Send tokens to contract: Invalid amount");
+            panic_with_error!(env, ContractError::InvalidTransferAmount);
+        }
+
+        let token_client = token_contract::Client::new(&env, &get_config(&env).token_info.address);
+        token_client.transfer(&sender, &contract, &amount);
+
+        env.events().publish(
+            (
+                "Send tokens to contract",
+                "Sent tokens to contract from account: {}, {}",
+            ),
+            (sender, contract),
+        );
+    }
+
+    fn add_to_whitelist(env: Env, sender: Address, to_add: Address) {
+        let mut config = get_config(&env);
+        if sender != config.admin {
+            log!(
+                &env,
+                "Vesting: Add to whitelist: Not authorized to add to whitelist"
+            );
+            panic_with_error!(env, ContractError::NotAuthorized);
+        }
+
+        config.whitelist.push_back(to_add.clone());
+        save_config(&env, &config);
+
+        env.events()
+            .publish(("Add to whitelist", "Added to whitelist: "), to_add);
+    }
+
+    fn remove_from_whitelist(env: Env, sender: Address, to_remove: Address) {
+        let mut config = get_config(&env);
+        if sender != config.admin {
+            log!(
+                &env,
+                "Vesting: Remove from whitelist: Not authorized to remove from whitelist"
+            );
+            panic_with_error!(env, ContractError::NotAuthorized);
+        }
+
+        if let Some(index) = config.whitelist.first_index_of(to_remove.clone()) {
+            config.whitelist.remove(index);
+        }
+
+        save_config(&env, &config);
+
+        env.events().publish(
+            ("Remove from whitelist", "Removed from whitelist: "),
+            to_remove,
+        );
+    }
+
+    fn query_config(env: Env) -> Result<Config, ContractError> {
+        let config = get_config(&env);
+
+        Ok(config)
+    }
+
+    fn query_balance(env: Env, address: Address) -> i128 {
+        let token_client = token_contract::Client::new(&env, &get_config(&env).token_info.address);
+        token_client.balance(&address)
+    }
+
+    fn query_vesting(env: Env, address: Address) -> Result<Curve, ContractError> {
+        let curve = get_vesting(&env, &address)?.curve;
+        Ok(curve)
```

Comment:
```suggestion
        get_vesting(&env, &address)?.curve
```

Reply:
kept the return type of this function as `Result`, so we can use `?`